File: gabes-tatenafas/backend/api/fuzzy-type2.php
<?php
/**
 * Interval Type-2 Fuzzy Logic endpoint — RÉEL.
 * Les entrees (pollution, humidite, temperature...) et les scores par zone sont
 * calcules depuis les vraies mesures api_readings. La reduction de type
 * (Karnik-Mendel) est un vrai algorithme deterministe.
 *   GET /backend/api/fuzzy-type2.php
 */
require_once __DIR__ . '/../lib/helpers.php';
require_once __DIR__ . '/../lib/auth.php';
require_once __DIR__ . '/../config/cities.php';
require_once __DIR__ . '/../lib/sci_status.php';

$me = auth_user();
if (!$me || !in_array($me['role'], ['admin'], true)) {
    json_response(['ok' => false, 'error' => 'admin_or_health_only'], 403);
}

function tri($x, $a, $b, $c) {
    if ($x <= $a || $x >= $c) return 0.0;
    if ($x == $b) return 1.0;
    return $x < $b ? ($x - $a) / max(1e-9, $b - $a) : ($c - $x) / max(1e-9, $c - $b);
}
function trap($x, $a, $b, $c, $d) {
    if ($x <= $a || $x >= $d) return 0.0;
    if ($x >= $b && $x <= $c) return 1.0;
    return $x < $b ? ($x - $a) / max(1e-9, $b - $a) : ($d - $x) / max(1e-9, $d - $c);
}
// Ensembles flous VOLONTAIREMENT chevauchants : a tout niveau de pollution
// [0-100], au moins deux ensembles sont actifs. Sinon (comme avant), quand
// seule LOW etait active, la reduction Karnik-Mendel renvoyait TOUJOURS le
// centroide de LOW (15) avec une bande=0 -> toutes les zones a AQI bas
// affichaient le meme 15/15/15/0. Avec chevauchement, le score varie
// continûment avec l'AQI reel et differencie donc chaque zone.
$sets = [
    'LOW'     => ['umf' => ['trap', 0, 0, 18, 45],    'lmf' => ['trap', 0, 0, 12, 35],    'color' => '#16a34a'],
    'MEDIUM'  => ['umf' => ['tri', 0, 40, 75],        'lmf' => ['tri', 12, 40, 65],       'color' => '#d97706'],
    'HIGH'    => ['umf' => ['tri', 40, 70, 100],      'lmf' => ['tri', 52, 70, 92],       'color' => '#dc2626'],
    'EXTREME' => ['umf' => ['trap', 68, 90, 100, 100],'lmf' => ['trap', 80, 94, 100, 100],'color' => '#7e0023'],
];
function eval_mf($def, $x) {
    if ($def[0] === 'tri') return tri($x, $def[1], $def[2], $def[3]);
    return trap($x, $def[1], $def[2], $def[3], $def[4]);
}
function km_fuzzy($sets, $pollution) {
    $centroids = ['LOW' => 15, 'MEDIUM' => 40, 'HIGH' => 70, 'EXTREME' => 90];
    $yl_num = 0; $yl_den = 0; $yr_num = 0; $yr_den = 0;
    foreach ($sets as $name => $d) {
        $u = eval_mf($d['umf'], $pollution); $l = eval_mf($d['lmf'], $pollution);
        $yl_num += $centroids[$name] * $u; $yl_den += $u;
        $yr_num += $centroids[$name] * $l; $yr_den += $l;
    }
    $yl = $yl_den > 0 ? $yl_num / $yl_den : 50;
    $yr = $yr_den > 0 ? $yr_num / $yr_den : 50;
    if ($yl > $yr) { $t = $yl; $yl = $yr; $yr = $t; }
    $score = ($yl + $yr) / 2; $band = $yr - $yl;
    $risk = $score < 30 ? 'low' : ($score < 55 ? 'moderate' : ($score < 80 ? 'high' : 'critical'));
    return [$score, $yl, $yr, $band, $risk];
}

$xs = [];
for ($x = 0; $x <= 100; $x += 2) $xs[] = $x;
$mf = [];
foreach ($sets as $name => $d) {
    $umf = []; $lmf = [];
    foreach ($xs as $x) { $umf[] = round(eval_mf($d['umf'], $x), 3); $lmf[] = round(eval_mf($d['lmf'], $x), 3); }
    $mf[] = ['set' => $name, 'color' => $d['color'], 'umf' => $umf, 'lmf' => $lmf];
}

$demo = false;
$pollution = 0; $vulnerability = 0; $symptom = 0; $alerts24 = 0;
$cityRows = [];
try {
    $pdo = db();
    $r = $pdo->query("SELECT final_aqi, final_humidity, final_temperature, final_pm25, final_pm10, final_so2 FROM api_readings WHERE final_aqi IS NOT NULL ORDER BY timestamp DESC LIMIT 1")->fetch();
    if ($r) {
        $pollution = min(100, (float)$r['final_aqi'] / 5.0);
        $vulnerability = min(100, max(0, ((((float)$r['final_humidity']) / 100) * 0.3 + ((((float)$r['final_temperature']) - 20) / 40) * 0.3 + (75000 / 150000) * 0.4) * 100));
        $symptom = min(100, max(0, ((((float)$r['final_pm25']) / 75) * 0.5 + (((float)$r['final_pm10']) / 150) * 0.3 + (((float)$r['final_so2']) / 100) * 0.2) * 100));
    } else {
        $demo = true;
    }
    try {
        $a = $pdo->query("SELECT COUNT(*) c FROM notifications WHERE created_at >= (NOW() - INTERVAL 24 HOUR)")->fetch();
        $alerts24 = (int)($a['c'] ?? 0);
    } catch (Throwable $e2) { $alerts24 = 0; }

    $cities = function_exists('gabes_cities') ? gabes_cities() : [];
    foreach ($cities as $zid => $c) {
        $st = $pdo->prepare("SELECT final_aqi a FROM api_readings WHERE city_id = ? AND final_aqi IS NOT NULL ORDER BY timestamp DESC LIMIT 1");
        $st->execute([(string)$zid]);
        $row = $st->fetch();
        $aqi = (float)($row['a'] ?? 0);
        if ($aqi <= 0) continue;
        $p = min(100, $aqi / 5.0);
        list($s, $yl, $yr, $band, $rl) = km_fuzzy($sets, $p);
        $cityRows[] = ['name' => $c['name_fr'] ?? ('Zone ' . $zid), 'name_ar' => $c['name_ar'] ?? '',
            'score' => round($s, 1), 'lower' => round($yl, 1), 'upper' => round($yr, 1),
            'band' => round($band, 1), 'risk' => $rl];
    }
} catch (Throwable $e) {
    $demo = true;
}

$degrees = [];
foreach ($sets as $name => $d) {
    $degrees[] = ['set' => $name, 'color' => $d['color'],
        'umf' => round(eval_mf($d['umf'], $pollution), 3),
        'lmf' => round(eval_mf($d['lmf'], $pollution), 3)];
}
list($score, $yl, $yr, $band, $risk) = km_fuzzy($sets, $pollution);

json_response([
    'ok' => true, 'demo' => $demo,
    'x' => $xs, 'mf' => $mf, 'degrees' => $degrees,
    'inputs' => ['pollution' => round($pollution, 1), 'vulnerability' => round($vulnerability, 1),
                 'symptom_severity' => round($symptom, 1), 'alerts_24h' => $alerts24],
    'score' => ['fuzzy_score_type2' => round($score, 1), 'uncertainty_lower' => round($yl, 1),
                'uncertainty_upper' => round($yr, 1), 'uncertainty_band' => round($band, 1), 'risk_level' => $risk],
    'cities' => $cityRows,
    'reference' => 'Mendel (2017), Springer',
]);

File: gabes-tatenafas/models/fuzzy_type2.py
"""PART 1 - Interval Type-2 Fuzzy Logic (Mamdani -> Type-2) with Karnik-Mendel.

Produces `fuzzy_score_type2` (the key ML/DL feature) plus its interval
uncertainty band. Pure NumPy - no heavy deps required.

Reference: Mendel (2017). Type-2 Fuzzy Systems. Springer.
"""
from __future__ import annotations
import numpy as np


def tri(x, a, b, c):
    x = np.asarray(x, dtype=float)
    left = (x - a) / (b - a + 1e-9)
    right = (c - x) / (c - b + 1e-9)
    return np.clip(np.minimum(left, right), 0, 1)


def trap(x, a, b, c, d):
    x = np.asarray(x, dtype=float)
    left = (x - a) / (b - a + 1e-9)
    right = (d - x) / (d - c + 1e-9)
    return np.clip(np.minimum(np.minimum(left, right), 1), 0, 1)


# Interval Type-2 membership functions (UMF, LMF) for every variable
SETS = {
    "LOW":     {"umf": ("trap", 0, 0, 20, 35),    "lmf": ("trap", 0, 0, 15, 28)},
    "MEDIUM":  {"umf": ("tri", 25, 50, 75),       "lmf": ("tri", 30, 50, 70)},
    "HIGH":    {"umf": ("trap", 60, 75, 90, 100), "lmf": ("trap", 65, 78, 88, 100)},
    "EXTREME": {"umf": ("trap", 80, 90, 100, 100),"lmf": ("trap", 85, 93, 100, 100)},
}
CENTROIDS = {"LOW": 15, "MEDIUM": 40, "HIGH": 70, "EXTREME": 90}


def _mf(defn, x):
    return tri(x, *defn[1:]) if defn[0] == "tri" else trap(x, *defn[1:])


def membership(value):
    """Return dict set -> (lmf_degree, umf_degree) for a scalar input."""
    out = {}
    for name, d in SETS.items():
        out[name] = (float(_mf(d["lmf"], value)), float(_mf(d["umf"], value)))
    return out


def karnik_mendel(firing):
    """Simplified KM type-reduction over centroids given interval firing.

    firing: dict set -> (lower, upper). Returns (y_l, y_r).
    """
    xs = np.array([CENTROIDS[k] for k in SETS])
    lo = np.array([firing[k][0] for k in SETS])
    hi = np.array([firing[k][1] for k in SETS])
    yl = float((xs * hi).sum() / (hi.sum() + 1e-9))  # left uses UMF (wider)
    yr = float((xs * lo).sum() / (lo.sum() + 1e-9))  # right uses LMF
    if yl > yr:
        yl, yr = yr, yl
    return yl, yr


def compute_inputs(row, population=75000):
    """Build the 4 fuzzy inputs from an api_readings row (dict-like)."""
    pollution = min(100.0, float(row.get("final_aqi", 0)) / 5.0)
    vulnerability = min(100.0, max(0.0, (
        (float(row.get("final_humidity", 50)) / 100) * 0.3 +
        ((float(row.get("final_temperature", 25)) - 20) / 40) * 0.3 +
        (population / 150000) * 0.4) * 100))
    symptom = min(100.0, max(0.0, (
        (float(row.get("final_pm25", 0)) / 75) * 0.5 +
        (float(row.get("final_pm10", 0)) / 150) * 0.3 +
        (float(row.get("final_so2", 0)) / 100) * 0.2) * 100))
    return pollution, vulnerability, symptom


def assess(pollution, vulnerability=50.0, symptom=50.0, alerts_24h=0):
    """Full Type-2 assessment. Returns dict ready for `fuzzy_assessments`."""
    # aggregate firing across inputs (rule weight via mean of inputs)
    firing = {}
    for k in SETS:
        lo = np.mean([membership(v)[k][0] for v in (pollution, vulnerability, symptom)])
        hi = np.mean([membership(v)[k][1] for v in (pollution, vulnerability, symptom)])
        firing[k] = (float(lo), float(hi))
    yl, yr = karnik_mendel(firing)
    score = (yl + yr) / 2
    band = yr - yl
    risk = ("low" if score < 30 else "moderate" if score < 55
            else "high" if score < 80 else "critical")
    return {
        "fuzzy_score_type2": round(score, 2),
        "uncertainty_lower": round(yl, 2),
        "uncertainty_upper": round(yr, 2),
        "uncertainty_band": round(band, 2),
        "risk_level": risk,
    }


if __name__ == "__main__":
    demo = assess(78, 62, 55, 4)
    print("Type-2 fuzzy assessment:", demo)
